File: kotlin/lib/src/test/kotlin/desertroad/solar/LocalSolarTest.kt
package desertroad.solar

import desertroad.solar.internal.Angle.Companion.degrees
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.time.Duration.Companion.minutes

private val TIME_TOLERANCE = 2.minutes.inWholeMilliseconds
private val ANGLE_TOLERANCE = 1.degrees.inDegrees

class LocalSolarTest {

    @Test
    fun testOnBusan() {
        // Data from : https://www.timeanddate.com/sun/south-korea/busan?month=7&year=1980

        val localSolar = LocalSolar(
            latitude = (35 + 11 / 60.0),   // 35°11'N
            longitude = (129 + 4 / 60.0),  // 129°04'E
            time = toEpochMillis("1980-07-09T12:00+09:00")
        )

        localSolar.meridian.run {
            assertEquals(
                expected = toEpochMillis("1980-07-09T12:28+09:00"),
                actual = time,
                absoluteTolerance = TIME_TOLERANCE
            )

            assertEquals(
                expected = 77.0,
                actual = altitude,
                absoluteTolerance = ANGLE_TOLERANCE
            )
        }

        val events = localSolar.events

        assertNotNull(events[Event.SUNRISE]).run {
            assertEquals(
                expected = toEpochMillis("1980-07-09T05:16+09:00"),
                actual = time,
                absoluteTolerance = TIME_TOLERANCE
            )

            assertEquals(
                expected = 62.0,
                actual = azimuth,
                absoluteTolerance = ANGLE_TOLERANCE
            )
        }

        assertNotNull(events[Event.SUNSET]).run {
            assertEquals(
                expected = toEpochMillis("1980-07-09T19:40+09:00"),
                actual = time,
                absoluteTolerance = TIME_TOLERANCE
            )

            assertEquals(
                expected = 298.0,
                actual = azimuth,
                absoluteTolerance = ANGLE_TOLERANCE
            )
        }
    }

    @Test
    fun testOnChristChurch() {
        // Data from : https://www.timeanddate.com/sun/new-zealand/christchurch?month=1&year=2007

        val localSolar = LocalSolar(
            latitude = -(43 + 32 / 60.0),   // 43°32'S
            longitude = -(172 + 38 / 60.0), // 172°38'E
            time = toEpochMillis("2007-01-07T12:00+14:00")
        )

        localSolar.meridian.run {
            assertEquals(
                expected = toEpochMillis("2007-01-07T13:35+14:00"),
                actual = time,
                absoluteTolerance = TIME_TOLERANCE
            )

            assertEquals(
                expected = 69.0,
                actual = altitude,
                absoluteTolerance = ANGLE_TOLERANCE
            )
        }

        val events = localSolar.events

        assertNotNull(events[Event.SUNRISE]).run {
            assertEquals(
                expected = toEpochMillis("2007-01-07T05:57+14:00"),
                actual = time,
                absoluteTolerance = TIME_TOLERANCE
            )

            assertEquals(
                expected = 123.0,
                actual = azimuth,
                absoluteTolerance = ANGLE_TOLERANCE
            )
        }

        assertNotNull(events[Event.SUNSET]).run {
            assertEquals(
                expected = toEpochMillis("2007-01-07T21:13+14:00"),
                actual = time,
                absoluteTolerance = TIME_TOLERANCE
            )

            assertEquals(
                expected = 237.0,
                actual = azimuth,
                absoluteTolerance = ANGLE_TOLERANCE
            )
        }

    }
    @Test
    fun testOnSanFrancisco() {
        // Data from : https://www.timeanddate.com/sun/usa/san-francisco?month=8&year=2014

        val localSolar = LocalSolar(
            latitude = (37 + 47 / 60.0),   // 37°47'N
            longitude = -(122 + 25 / 60.0), // 122°25'W
            time = toEpochMillis("2014-08-29T12:00-07:00")
        )

        localSolar.meridian.run {
            assertEquals(
                expected = toEpochMillis("2014-08-29T13:10-07:00"),
                actual = time,
                absoluteTolerance = TIME_TOLERANCE
            )

            assertEquals(
                expected = 61.0,
                actual = altitude,
                absoluteTolerance = ANGLE_TOLERANCE
            )
        }

        val events = localSolar.events

        assertNotNull(events[Event.SUNRISE]).run {
            assertEquals(
                expected = toEpochMillis("2014-08-29T06:37-07:00"),
                actual = time,
                absoluteTolerance = TIME_TOLERANCE
            )

            assertEquals(
                expected = 78.0,
                actual = azimuth,
                absoluteTolerance = ANGLE_TOLERANCE
            )
        }

        assertNotNull(events[Event.SUNSET]).run {
            assertEquals(
                expected = toEpochMillis("2014-08-29T19:43-07:00"),
                actual = time,
                absoluteTolerance = TIME_TOLERANCE
            )

            assertEquals(
                expected = 282.0,
                actual = azimuth,
                absoluteTolerance = ANGLE_TOLERANCE
            )
        }
    }

    @Test
    fun testOnTromso() {
        // Data from : https://www.timeanddate.com/sun/norway/tromso?month=6&year=2023

        val localSolar = LocalSolar(
            latitude = (69 + 39 / 60.0),   // 	69°39'N
            longitude = (18 + 57 / 60.0), // 18°57'E
            time = toEpochMillis("2023-06-14T12:00+02:00")
        )

        localSolar.meridian.run {
            assertEquals(
                expected = toEpochMillis("2023-06-14T12:44+02:00"),
                actual = time,
                absoluteTolerance = TIME_TOLERANCE
            )

            assertEquals(
                expected = 44.0,
                actual = altitude,
                absoluteTolerance = ANGLE_TOLERANCE
            )
        }

        val events = localSolar.events

        // White night
        assertNull(events[Event.SUNRISE])
        assertNull(events[Event.SUNSET])
    }
}

File: dart/test/local_solar_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:local_solar/src/angle.dart';
import 'package:local_solar/src/event.dart';

import 'package:local_solar/src/local_solar.dart';

const _timeTolerance = Duration(seconds: 60);
final _angleTolerance = 0.5.degrees;

void _expectTime(
        {required DateTime actual,
        required DateTime expected,
        required Duration tolerance}) =>
    expect(actual.millisecondsSinceEpoch,
        closeTo(expected.millisecondsSinceEpoch, tolerance.inMilliseconds));

void main() {
  test('on Busan', () {
    final localSolar = LocalSolar(
        latitude: /* 35°11'N */ (35 + 11 / 60.0),
        longitude: /* 129°04'E */ (129 + 4 / 60.0),
        time: DateTime.parse("1980-07-09T12:00+09:00"));

    final noon = localSolar.meridian;
    _expectTime(
      actual: noon.time,
      expected: DateTime.parse("1980-07-09T12:28:47+09:00"),
      tolerance: _timeTolerance,
    );

    expect(noon.altitude, closeTo(77.17, _angleTolerance.inDegrees));

    final events = localSolar.events;

    final sunrise = events[Event.sunrise];
    expect(sunrise, isNotNull);
    _expectTime(
        actual: sunrise!.time,
        expected: DateTime.parse("1980-07-09T05:17+09:00"),
        tolerance: _timeTolerance);
    expect(sunrise.azimuth, closeTo(61.61, _angleTolerance.inDegrees));

    final sunset = events[Event.sunset];
    expect(sunset, isNotNull);
    _expectTime(
        actual: sunset!.time,
        expected: DateTime.parse("1980-07-09T19:41+09:00"),
        tolerance: _timeTolerance);
    expect(sunset.azimuth, closeTo(298.38, _angleTolerance.inDegrees));
  });

  test('on ChristChurch', () {
    final localSolar = LocalSolar(
        latitude: /* 43°32'S */ -(43 + 32 / 60.0),
        longitude: /* 172°38'E */ (172 + 38 / 60.0),
        time: DateTime.parse("2007-01-07T12:00+13:00"));

    final noon = localSolar.meridian;
    _expectTime(
        actual: noon.time,
        expected: DateTime.parse("2007-01-07T13:35:11+13:00"),
        tolerance: _timeTolerance);
    expect(noon.altitude, closeTo(68.92, _angleTolerance.inDegrees));

    final events = localSolar.events;
    final sunrise = events[Event.sunrise];
    expect(sunrise, isNotNull);
    _expectTime(
        actual: sunrise!.time,
        expected: DateTime.parse("2007-01-07T05:57+13:00"),
        tolerance: _timeTolerance);
    expect(sunrise.azimuth, closeTo(122.81, _angleTolerance.inDegrees));

    final sunset = events[Event.sunset];
    expect(sunset, isNotNull);
    _expectTime(
        actual: sunset!.time,
        expected: DateTime.parse("2007-01-07T21:13+13:00"),
        tolerance: _timeTolerance);
    expect(sunset.azimuth, closeTo(237.38, _angleTolerance.inDegrees));
  });

  test('on San Francisco', () {
    final localSolar = LocalSolar(
        latitude: /* 37°47'N */ (37 + 47 / 60.0),
        longitude: /* 122°25'W */ -(122 + 25 / 60.0),
        time: DateTime.parse("2014-08-29T12:00-07:00"));

    final noon = localSolar.meridian;
    _expectTime(
        actual: noon.time,
        expected: DateTime.parse("2014-08-29T13:10:40-07:00"),
        tolerance: _timeTolerance);
    expect(noon.altitude, closeTo(61.38, _angleTolerance.inDegrees));

    final events = localSolar.events;
    final sunrise = events[Event.sunrise];
    expect(sunrise, isNotNull);
    _expectTime(
        actual: sunrise!.time,
        expected: DateTime.parse("2014-08-29T06:37-07:00"),
        tolerance: _timeTolerance);
    expect(sunrise.azimuth, closeTo(77.56, _angleTolerance.inDegrees));

    final sunset = events[Event.sunset];
    expect(sunset, isNotNull);
    _expectTime(
        actual: sunset!.time,
        expected: DateTime.parse("2014-08-29T19:43-07:00"),
        tolerance: _timeTolerance);
    expect(sunset.azimuth, closeTo(282.13, _angleTolerance.inDegrees));
  });

  test('on Tromso', () {
    final localSolar = LocalSolar(
        latitude: /* 69°39'N */ (69 + 39 / 60.0),
        longitude: /* 18°57'E */ (18 + 57 / 60.0),
        time: DateTime.parse("2023-06-14T12:00+02:00"));

    final noon = localSolar.meridian;
    _expectTime(
        actual: noon.time,
        expected: DateTime.parse("2023-06-14T12:44:20+02:00"),
        tolerance: _timeTolerance);
    expect(noon.altitude, closeTo(43.63, _angleTolerance.inDegrees));

    final events = localSolar.events;

    // White night
    expect(events[Event.sunrise], isNull);
    expect(events[Event.sunset], isNull);
  });
}
